chalk-integration/src/lowering/env.rs

```diff
@@ -0,0 +1,269 @@
+use chalk_ir::interner::HasInterner;
+use chalk_ir::{
+    self, AdtId, BoundVar, ClosureId, DebruijnIndex, FnDefId, OpaqueTyId, TraitId, VariableKinds,
+};
+use chalk_ir::{cast::Cast, ForeignDefId, WithKind};
+use chalk_parse::ast::*;
+use chalk_solve::rust_ir::AssociatedTyValueId;
+use std::collections::BTreeMap;
+
+use crate::error::RustIrError;
+use crate::interner::ChalkIr;
+use crate::{Identifier as Ident, TypeKind};
+
+pub type AdtIds = BTreeMap<Ident, chalk_ir::AdtId<ChalkIr>>;
+pub type FnDefIds = BTreeMap<Ident, chalk_ir::FnDefId<ChalkIr>>;
+pub type ClosureIds = BTreeMap<Ident, chalk_ir::ClosureId<ChalkIr>>;
+pub type TraitIds = BTreeMap<Ident, chalk_ir::TraitId<ChalkIr>>;
+pub type OpaqueTyIds = BTreeMap<Ident, chalk_ir::OpaqueTyId<ChalkIr>>;
+pub type AdtKinds = BTreeMap<chalk_ir::AdtId<ChalkIr>, TypeKind>;
+pub type FnDefKinds = BTreeMap<chalk_ir::FnDefId<ChalkIr>, TypeKind>;
+pub type ClosureKinds = BTreeMap<chalk_ir::ClosureId<ChalkIr>, TypeKind>;
+pub type TraitKinds = BTreeMap<chalk_ir::TraitId<ChalkIr>, TypeKind>;
+pub type AutoTraits = BTreeMap<chalk_ir::TraitId<ChalkIr>, bool>;
+pub type OpaqueTyKinds = BTreeMap<chalk_ir::OpaqueTyId<ChalkIr>, TypeKind>;
+pub type AssociatedTyLookups = BTreeMap<(chalk_ir::TraitId<ChalkIr>, Ident), AssociatedTyLookup>;
+pub type AssociatedTyValueIds =
+    BTreeMap<(chalk_ir::ImplId<ChalkIr>, Ident), AssociatedTyValueId<ChalkIr>>;
+pub type ForeignIds = BTreeMap<Ident, chalk_ir::ForeignDefId<ChalkIr>>;
+
+pub type ParameterMap = BTreeMap<Ident, chalk_ir::WithKind<ChalkIr, BoundVar>>;
+
+pub type LowerResult<T> = Result<T, RustIrError>;
+
+#[derive(Clone, Debug)]
+pub struct Env<'k> {
+    pub adt_ids: &'k AdtIds,
+    pub adt_kinds: &'k AdtKinds,
+    pub fn_def_ids: &'k FnDefIds,
+    pub fn_def_kinds: &'k FnDefKinds,
+    pub closure_ids: &'k ClosureIds,
+    pub closure_kinds: &'k ClosureKinds,
+    pub trait_ids: &'k TraitIds,
+    pub trait_kinds: &'k TraitKinds,
+    pub opaque_ty_ids: &'k OpaqueTyIds,
+    pub opaque_ty_kinds: &'k OpaqueTyKinds,
+    pub associated_ty_lookups: &'k AssociatedTyLookups,
+    pub auto_traits: &'k AutoTraits,
+    pub foreign_ty_ids: &'k ForeignIds,
+    /// GenericArg identifiers are used as keys, therefore
+    /// all identifiers in an environment must be unique (no shadowing).
+    pub parameter_map: ParameterMap,
+}
+
+/// Information about an associated type **declaration** (i.e., an
+/// `AssociatedTyDatum`). This information is gathered in the first
+/// phase of creating the Rust IR and is then later used to lookup the
+/// "id" of an associated type.
+///
+/// ```ignore
+/// trait Foo {
+///     type Bar<'a>; // <-- associated type declaration
+///          // ----
+///          // |
+///          // addl_variable_kinds
+/// }
+/// ```
+#[derive(Debug, PartialEq, Eq)]
+pub struct AssociatedTyLookup {
+    pub id: chalk_ir::AssocTypeId<ChalkIr>,
+    pub addl_variable_kinds: Vec<chalk_ir::VariableKind<ChalkIr>>,
+}
+
+pub enum TypeLookup<'k> {
```

Comment:
I would argue this isn't *quite* `TypeLookup` since `Trait` isn't a `type`, but eh. This is probably fine.